src/Workspaces/Core/Portable/Execution/Serialization/Serializer_HierarchicalChecksumObject.cs

```diff
@@ -0,0 +1,163 @@
+
+            var kind = reader.ReadString();
+            Contract.ThrowIfFalse(kind == SolutionChecksumObject.Name);
+
+            var checksum = Checksum.ReadFrom(reader);
+
+            var info = Checksum.ReadFrom(reader);
+            var projects = DeserializeChecksumCollection(reader, cancellationToken);
+
+            var snapshotId = new SolutionChecksumObject(this, info, projects);
+            Contract.ThrowIfFalse(checksum.Equals(snapshotId.Checksum));
+
+            return snapshotId;
+        }
+
+        public async Task SerializeAsync(ProjectChecksumObject snapshotId, ObjectWriter writer, CancellationToken cancellationToken)
+        {
+            writer.WriteString(snapshotId.Kind);
+            snapshotId.Checksum.WriteTo(writer);
+
+            snapshotId.Info.WriteTo(writer);
+            snapshotId.CompilationOptions.WriteTo(writer);
+            snapshotId.ParseOptions.WriteTo(writer);
+
+            await snapshotId.Documents.WriteToAsync(writer, cancellationToken).ConfigureAwait(false);
+
+            await snapshotId.ProjectReferences.WriteToAsync(writer, cancellationToken).ConfigureAwait(false);
+            await snapshotId.MetadataReferences.WriteToAsync(writer, cancellationToken).ConfigureAwait(false);
+            await snapshotId.AnalyzerReferences.WriteToAsync(writer, cancellationToken).ConfigureAwait(false);
+
+            await snapshotId.AdditionalDocuments.WriteToAsync(writer, cancellationToken).ConfigureAwait(false);
+        }
+
+        private ProjectChecksumObject DeserializeProjectChecksumObject(ObjectReader reader, CancellationToken cancellationToken)
+        {
+            var kind = reader.ReadString();
+            Contract.ThrowIfFalse(kind == ProjectChecksumObject.Name);
+
+            var checksum = Checksum.ReadFrom(reader);
+
+            var info = Checksum.ReadFrom(reader);
+            var compilationOptions = Checksum.ReadFrom(reader);
+            var parseOptions = Checksum.ReadFrom(reader);
+
+            var documents = DeserializeChecksumCollection(reader, cancellationToken);
+
+            var projectReferences = DeserializeChecksumCollection(reader, cancellationToken);
+            var metadataReferences = DeserializeChecksumCollection(reader, cancellationToken);
+            var analyzerReferences = DeserializeChecksumCollection(reader, cancellationToken);
+
+            var additionalDocuments = DeserializeChecksumCollection(reader, cancellationToken);
+
+            var snapshotId = new ProjectChecksumObject(
+                this,
+                info, compilationOptions, parseOptions, documents,
+                projectReferences, metadataReferences, analyzerReferences, additionalDocuments);
+            Contract.ThrowIfFalse(checksum.Equals(snapshotId.Checksum));
+
+            return snapshotId;
+        }
+
+        public void Serialize(DocumentChecksumObject snapshotId, ObjectWriter writer, CancellationToken cancellationToken)
```

Comment:
Can you keep the serialize and deserialize names consistent with each other?  i.e. SerializeDocumentChecksumObject